cmake: revert CUDA arch default to 52, 61 if f16

JohannesGaessler opened the pull request:
In https://github.com/ggerganov/llama.cpp/pull/1913 I changed cmake to use `native` as the CUDA arch (like the Makefile does). However, this seems to cause compilation issues for some users which seems to be related to cmake version, see https://github.com/ggerganov/llama.cpp/pull/1946. This PR restores the default cmake CUDA arch to `compute_52`. If `LLAMA_CUDA_DMMV_F16` is set the arch is instead set to `compute_61`. [When I tested it using an RTX 3090](https://github.com/ggerganov/llama.cpp/pull/1946#issuecomment-1600356433), I was not able to discern a performance difference from using `native` over `compute_52`. This PR does **not** fix the inconsistent behavior between make and cmake.

JohannesGaessler committed:
cmake: revert CUDA arch default to 52, 61 if f16

howard0su reviewed (approved):
Looks good to me.

slaren commented:
The `CUDA_SELECT_NVCC_ARCH_FLAGS` can be removed, it doesn't really work. I think it would be better to explicitly set `LLAMA_CUDA_ARCH` to 52 rather to leave it to nvcc, cmake will not pass any `-arch` flags if you set it to `Auto`.

slaren commented:
We can avoid setting `CUDA_ARCHITECTURES` for each target individually with `CMAKE_CUDA_ARCHITECTURES`. It will also give a way for users to use a different value by setting the `CUDAARCHS` environment variable.
```cmake
if(NOT DEFINED CMAKE_CUDA_ARCHITECTURES)
  set(CMAKE_CUDA_ARCHITECTURES 75)
endif()
```
https://cmake.org/cmake/help/latest/variable/CMAKE_CUDA_ARCHITECTURES.html

JohannesGaessler commented:
Apparently you need to set `CMAKE_CUDA_ARCHITECTURES` before you set the target so I moved it upwards to the check where it's looking for cuBLAS. I also added an info print that just tells the user which architecture is being used instead of telling them that the architecture is being configured.
